### DISCUSSION TITLE
criteriaMode: formLevel

### DISCUSSION THREAD

**Discussion by bluebill1049** [react-hook-form/react-hook-form#6472] [RFC]
## Context

React Hook Form focus on performance and user experience, which is a result of why the lib has quite precise on when to re-render your components. For example, if the input contains an error, we will trigger exactly one re-render to info the component to update, and exactly one re-render when the input's error gets resolved (instead of spamming the re-render). This is one of the reasons hook form is unique and more performant than most of the form libraries in the industry. The also reflects on form state update. However, we would like to expand that option to users, and allow users to have validation at the form level. So what are some of the impacts on such config:

### Props

- User will be able to watch the entire form level validation errors
- Schema errors will be returned as a whole without any optimization or omit
- Works both with build-in and schema validation
- There are cases where custom resolver can be skipped

### Cons

- If such config is turned on, the form will be less performant, which means each keystroke will hitting the validation rules and trigger more re-render to make sure all errors are presented for the entire form.
- More code to maintain and package size
- validation strategy `mode` & `reValidateMode` will be ignored

## Proposal

```tsx
useForm({
  criteriaMode: 'formLevel', // other options: firstError, all
})
```

## Related issues

https://github.com/react-hook-form/react-hook-form/issues/4356
https://github.com/react-hook-form/resolvers/issues/243

cc @jorisre @kotarella1110 @barrymay

**Comment by wdfinch** [comment]
Sounds good to me. Two questions:

Would these errors be placed in the `formState.error` object?
How would `trigger` be affected by this?

**Reply by bluebill1049** [reply]
> Would these errors be placed in the formState.error object?

Yes,

> How would trigger be affected by this?

will always return the form errors

**Reply by wdfinch** [reply]
Ok nice. Ya sounds good to me.

**Reply by bluebill1049** [reply]
actually `trigger` shouldn't be affected at all. the functionality stays as it is.

**Comment by bluebill1049** [comment]
Question: do we need to flush out all the form errors when the inputs are mounted or until the user interacts with the inputs? 

The difference is pretty much whether we invoke trigger() instead of the lib at `useEffect`, or actually give users the option and when and what to fire the trigger at their app level.

**Reply by wdfinch** [reply]
I think when the user interacts with the inputs. This is how regular errors work. We add to the errors object once the user interacts.

**Reply by wdfinch** [reply]
I made a feature request for showing the errors on the same cycle as isValid but this is a separate use case and feature in my mind.

**Comment by bluebill1049** [comment]
With `formLevel` option, this will ignore the `mode/reValdiateMode` option as well, because all errors will be returned by the hook.

**Reply by bluebill1049** [reply]
updated in the con section as well.

**Comment by barrymay** [comment]
I'm a little confused by this one, although I think I like the idea - can you explain how you would document/explain formLevel vs the existing firstError vs all?   (a table comparison might help, as in what setting has what behavior)

If it's what I think it is, I can see this being a commonly desired behavior and one the devs would sacrifice perf for vs the clearer user behavior.  I feel I've likely made my forms work this way at times using mode and revalidateMode but having this more "built in" would be nice.

**Reply by wdfinch** [reply]
The issue now is that only errors that corespond to inputs/fields are passed into `formState.errors`. This means if we want to do something like validate that a field array has at least one element, we cannot display a clear error message to the user without additional work. The goal of this would be to not omit anything from `formState.errors` and stay true to what our validation is actually returning.

**Reply by bluebill1049** [reply]
Its basically asking for hook form to return every single error on your entire form. Skip all validation strategy and dump all the errors to the app use.

**Reply by barrymay** [reply]
got it, thanks - that's what I thought it was.  I agree it's likely a common desire and an issue devs may have with the default RHF behaviors.

**Reply by bluebill1049** [reply]
The more I think about this, the more concern I have with this approach and its API design. if this is an issue related to the field array, perhaps the emphasis should be on the field array. eg each field array action triggers the validation schema and return the field array named errors.

```tsx
useFieldArray({ name: 'fieldArray' })

append({ whatever: 'test' })

trigger('fieldArray') // internal 

// then re-render with any errors if occurred
```

I think validation strategy is one of the strengths of hook form and I don't want to lose that.

**Reply by barrymay** [reply]
I agree the API design of this needs more thought, and it's to an extent why I mentioned my note about a form-wide validation fn below - at minimum I could see a change here that might introduce a new criteriaMode (or whatever its called) should throw a dev-time type error on flags that are not compatible with that mode, so there's less confusion.

**Reply by bluebill1049** [reply]
I will play around with the above solution and keep you in the loop @barrymay.

**Reply by bluebill1049** [reply]
100% agree on @barrymay, I think the best is to avoid the new `crtieraMode` if we can. It's getting more confusing the more I think about it.

**Reply by wdfinch** [reply]
Sounds good. Aside from field arrays which are my main use case, I could also see nested objects having some sort of validation. For example, if we have form values such as:

```
{
  hoursOfOperation: {
    mon: {
      openTime: string,
      closeTime: string
    },
    tues: {
      openTime: string,
      closeTime: string
    }
  }
}
```

I could see the user wanting to do some validation on the containers. For example, open time must come **before** close time and they want to validate each day object rather than the individual fields. So the error would look like this:

```
{
  hoursOfOperation: {
    mon: {
      message: 'opening time must come before close time',
      type: 'object.example'
    }
  }
}
```

You could do this one at the field level too, but just an example. In other words, I don't think we should restrict it to just field arrays.

**Reply by bluebill1049** [reply]
That's would be difficult, in that case it's easier to just do a `trigger` with onChange.

**Reply by wdfinch** [reply]
Ok. I guess if we call trigger every time, then a side effect is the error object would be filled with errors immediately then before the user touched the input right?

**Reply by bluebill1049** [reply]
can you provide a codesandbox? i would like to verify an idea which i had. i can test out tonight.

**Reply by wdfinch** [reply]
Sure what do you want to see in the sandbox?

**Reply by bluebill1049** [reply]
this example: https://github.com/react-hook-form/react-hook-form/discussions/6472#discussioncomment-1515283

**Reply by bluebill1049** [reply]
I can summarise the (DX) issue into two categories: 

- parent field errors (schema) 
- field array parent errors (schema) 

will need to try to unlock the first improvement first before going to field array

**Reply by wdfinch** [reply]
Sure. I'll work on it tonight.

**Reply by wdfinch** [reply]
@bluebill1049 here you go. This is interesting, it seems like object errors are actually reported in `formState.errors` as expected. In the example, I run some validation on an object that contains fields and the error gets reported correctly with a path that is not a path to a rendered input. Not sure if this is a bug or the intended behavior. So maybe the issue here is just with field arrays then? 

https://codesandbox.io/s/distracted-elbakyan-epkn8?file=/src/App.js

**Comment by wdfinch** [comment]
I'm not sure how this is a performance hit. The reason being is that we decide what validation we want to perform. All this is really doing is giving us access to errors that are already being computed by our validation (to get `formState.isValid` right).

Perhaps with built-in validation, it's more expensive if there are some optimizations. However, with schema validation, it is still being run the way we decide each time. For example, the user can use the `abortEarly` option in `joi` to stop after the first validation failure.

If anything this seems like it would be faster because we don't remove parse the schema result object and remove non-field errors. As mentioned above we can return it closer to how it is.

**Reply by bluebill1049** [reply]
It will have a performance impact, each keystroke will validate schema and trigger re-render according.

**Reply by barrymay** [reply]
I agree it's may likely be a performance hit, and the length to which it is is likely dependent on the implementation.  But as with anything UI and perf, it's about how much is really perceivable by the user to be noticeable.  If typing became a concern, you could also consider debouncing the validation if the user is typing fast, or only revalidate on a full field change commit.

@bluebill1049 I have another idea related to this which you may or may not have thought of/heard of - what about instead of needing to re-run many validation rules, maybe we consider allowing users to define a single form-wide validation function which returns essentially an error object, and RHF then applies it to its internal error mapping. 

That way if the error of one field has many side effects that may affect other field validations, or if users want to short-circuit the validation of a section for perf/logic reasons, they can?  (especially when it's a form with a long array list, this could be very handy)

**Reply by bluebill1049** [reply]
thanks, @barrymay that would solve the problem for in build validation field array, at the same time I try to avoid more APIs for the library if we can. for build in validation, the easier way is just at the `useEffect` and set an error.

**Reply by barrymay** [reply]
that's very fair, maybe this is just put on the RHF website as a implementation example (if one's not there already for this case)

**Reply by bluebill1049** [reply]
Will do 👍

**Reply by bluebill1049** [reply]
https://github.com/react-hook-form/documentation/issues/680

**Reply by wdfinch** [reply]
But we have the `mode` option on `useForm`. Couldn't the user just set this to `onSubmit` or something other than `onChange`? To me, this feature isn't about when we run validation but rather what errors get included in the `formState.errors` object

**Comment by wdfinch** [comment]
One question I have is should we place this under `criteriaMode`? To me, it seems like this controls something different.

For example, we could have multiple form-level errors. 

Ex: we want to make sure a field array that consists of a single text input has at least 4 inputs **and** all the text in the inputs are unique. With the `formLevel` option would it behave like `firstError` or `all`? In other words, would we get back 1 or 2 assuming errors both validation rules fail? 

To me, this seems like something that either should be set by default or a separate option. Again given my performance note above, I'm not sure how adding this would affect the performance. For example, if you don't care about form-level validation, then don't add any validation rules. Otherwise, we are already doing the work to compute it so I don't see how giving us the error message adds much more work.

**Reply by bluebill1049** [reply]
what's your suggestion?

**Reply by wdfinch** [reply]
I think this could just be the default behavior. If we write logic to validate at form level (a field array has 1 element), then I don't see how it hurts to include this information in the `formState.errors` object. In other words, if you don't want the validation, then don't write it. We are just giving people more information rather than omitting error messages. Whether the user wants to display the message is up to them. If anything it makes the logic easier because you don't need to parse the result of the schema validation and remove errors that don't relate to a field. Plus we still run the form level validation to compute `formState.isValid`.

Regarding performance, see my note above. I don't see why this needs to run on each keystroke and the user can't choose the mode option. This change is about not removing errors from the errors object rather than when we run the validation.

**Comment by wdfinch** [comment]
I guess the way I see this working, is we simply don't omit paths from `formState.errors`. For example, if we validate our form in `joi` and I require that a field array must have at least 1 element, then I get an error object that looks somewhat like this after running `scehma.validate(getValues)`:
```
{
  context: {limit: 1, value: Array(0), label: 'myFieldArray', key: 'myFieldArray'}
  message: "\"myFieldArray\" must contain at least 2 items"
  path: ['myFieldArray']
  type: "array.min"
}
```

So therefore what I would like to see is in my `formState.errors` object is:

```
{
  myFieldArray: {
    message: "\"myFieldArray\" must contain at least 2 items",
    type: "array.min"
  }
}
```

Whether we use `criteriaMode` to show multiple errors or a single error is independent of this.

**Reply by bluebill1049** [reply]
For formlevel we are basically calling trigger on each keystroke, similar to most of controlled form validation lib. Just keep flush out all the schema errors.

**Reply by wdfinch** [reply]
Gotcha. See notes above regarding this.

**Comment by mimameid** [comment]
I'm not quite sure if it is related to this topic, but since I got linked to here from [6858](https://github.com/react-hook-form/react-hook-form/discussions/6858) and I see that `fieldArrays` are also part of this I wanted to add my 2 Cents here.

After digging a bit deeper into my confusion about [6858](https://github.com/react-hook-form/react-hook-form/discussions/6858), I figured, that validation is working differently if the input is field level. When I compare it with field arrays: In both validation does run as defined in `mode`, though for `fieldArrays` the errors aren't reflected in the `formState` immediately, but only after submit. While this works for field level input. I can't find a documentation about this behavior. You can find an example in the link.

Is this related? If so, how come the validation is run anyway, but the error is not stored/reflected?

**Reply by wdfinch** [reply]
That's just the way it works. `formState.isValid` is based on the true result of the validation. So in schema validation, this means  `formState.isValid = schema.validate(formValues).isValid`. Not the exact code, but this is the idea. 

The errors object however only shows errors that relate to user-level input fields. So field array errors are not included because a field array is not a user-level input. The fields within a field array `myFieldArray.0.name` are user-level inputs, and errors for these are added to `formState.errors`.

The goal of this feature is to change this so we can get full access to all the errors whether they are related to a user input or not.

**Comment by bluebill1049** [comment]
here is an improved version: https://codesandbox.io/s/react-hook-form-fieldsarray-yup-validation-forked-6f6ih?file=/src/src/useFieldArray.ts

- The library will still **respect** the validation strategy (mode), we will not pass full errors to the form.
- useFieldArray now validate the entire field array with schema validation parent node, so each action will report if there is a group error
-  Checkbox/radio or group inputs validation will be reported

**Reply by wdfinch** [reply]
Looks great! Nice job. I think if we could just get field array errors like in the sandbox then this would solve the problem for almost all the users (including myself). Thanks for doing this.

**Reply by franksimon90** [reply]
@bluebill1049 the `trigger("*question*")` on line 51 is intended, or it's just a typo? Current behavior for me is still confusing, if you add `n` fields, and focus on some `k`, start getting errors for `k...n` fields, everything bellow is valid. I managed to mark all the fields as invalid using controlled components [here](https://codesandbox.io/s/react-hook-form-usefieldarray-template-forked-ng4eq4?file=/src/index.js), but I'm pretty sure that the approach that I took on that sandbox is wrong, any suggestions?
